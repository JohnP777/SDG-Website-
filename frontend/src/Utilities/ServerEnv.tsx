export const getUrl = () => {
    if (process.env.REACT_APP_BACKEND_URL) {
        return process.env.REACT_APP_BACKEND_URL;
    }
    return 'http://127.0.0.1:8000/'
}