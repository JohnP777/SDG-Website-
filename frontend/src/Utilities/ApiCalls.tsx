import { getUrl } from "./ServerEnv";

export const apiCallPost = async (path: string, body: any, isAuthed: boolean ) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'POST',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
      body: JSON.stringify(body)
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallGet = async (path: string, isAuthed: boolean, headersExtra?: any) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'GET',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : '',
        ...headersExtra
      },
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallPatch = async (path: string, body: object, isAuthed: boolean) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'PATCH',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
      body: JSON.stringify(body)
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallDelete = async (path: string, isAuthed: boolean) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'DELETE',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallPut = async (path: string, body: any, isAuthed: boolean) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'PUT',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
      body: JSON.stringify(body)
    });
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error);
  }
};
