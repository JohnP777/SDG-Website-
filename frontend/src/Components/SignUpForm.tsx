import React, { useState } from 'react';
import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import Divider from '@mui/material/Divider';
import InputField from './LoginInputField';
import { Alert, Typography } from '@mui/material';
import { apiCallPost } from '../Utilities/ApiCalls';
import { useNavigate } from 'react-router-dom';
import { GoogleLogin } from '@react-oauth/google';
import { useGoogleLogin } from './GoogleLogin';
import { Checkbox, FormControlLabel } from '@mui/material';

const SignUpForm = () => {

  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [revealed, setRevealed] = useState<'password' | 'text'>('password');
  const [email, setEmail] = useState('');
  const [phoneNumber, setPhoneNumber] = useState('');
  const [confirmPassword, setConfirmPassword] = useState('');

  const [errorMessage, setErrorMessage] = useState<string[]>([]);

  const navigate = useNavigate();

  const { handleGoogleLoginSuccess, handleGoogleLoginError } = useGoogleLogin();
  
  // API Call Here
  const signup = async () => {
    const data = await apiCallPost('api/auth/pending-register/', { username, email, password1: password, password2: confirmPassword, mobile: phoneNumber }, false);
    if (data.token) {
      navigate('/confirmation-pin', {
        state: {
          username,
          email,
          password1: password,
          password2: confirmPassword,
          mobile: phoneNumber,
          initialToken: data.token
        }}
      )
    } else {
      const errors: string[] = [];
      for (let field in data) {
        if (field !== 'statusCode') {
          if (field === 'non_field_errors') {
            errors.push(data[field][0]);
          } else {
            errors.push(field.toUpperCase() + ': ' + data[field][0]);
          }
        }
      }
      setErrorMessage(errors);
    }
  };

  // Show Password
  const handleToggle = (event: React.ChangeEvent<HTMLInputElement>) => {
    setRevealed(event.target.checked ? 'text' : 'password');
  };

  return (
    <Box sx={{
      boxSizing: 'border-box',
      width: { xs: '100%', sm: '50%' },
      maxWidth: '100%',
      padding: 3, 
      backgroundColor: 'white', 
      display: 'flex', 
      flexDirection: 'column',
      alignItems: 'center',
      borderRadius: { xs: '0 0 8px 8px', sm: '0 8px 8px 0' }
    }}>
      <Typography style={{ fontSize: '1.6rem', fontWeight: 'bold', marginBottom: '20px', textAlign: 'center' }}>SDG Knowledge Sign Up</Typography>

      {errorMessage.length > 0 && 
        (<Alert severity='error' sx={{ marginBottom: '15px' }}>
          {errorMessage.map((error, index) => (
            <Typography key={index} sx={{ fontSize: '0.85rem' }}>
              {error}
            </Typography>
          ))}
        </Alert>)
      }

      <Box sx={{ 
        display: 'flex', 
        justifyContent: 'space-around',
        alignItems: 'center', 
        flexDirection: 'column',
        gap: '10px',
        width: '80%'
      }}>
        <InputField 
          type='text'
          name='username' 
          placeholder='Username'
          value={username}
          onChange={(e) => setUsername(e.target.value)}
        />

        <InputField 
          type='text'
          name='contact' 
          placeholder='Email'
          value={email}
          onChange={(e) => setEmail(e.target.value)}
        />

        <InputField 
          type='text'
          name='phonenumber' 
          placeholder='Phone Number (optional)'
          value={phoneNumber}
          onChange={(e) => setPhoneNumber(e.target.value)}
        />
        <InputField 
          type={revealed}
          name='password' 
          placeholder='Password'
          value={password}
          onChange={(e) => setPassword(e.target.value)}
        />
        <InputField 
          type={revealed}
          name='confirmpassword' 
          placeholder='Confirm Password'
          value={confirmPassword}
          onChange={(e) => setConfirmPassword(e.target.value)}
        />
        <FormControlLabel
          sx={{ m: 0, '& .MuiFormControlLabel-label': { fontSize: '0.75rem' } }}
          control={
            <Checkbox
              size='small'
              checked={revealed === 'text'}
              onChange={handleToggle}
            />
          }
          label='Show Password'
        />

        <Button
          variant='contained'
          size='small' 
          sx={{
            backgroundColor: '#000000',
            width: '85%',
            fontSize: '10px',
            textTransform: 'none',
            boxShadow: 'none'
          }}
          onClick={signup}
        >
          Sign Up
        </Button>
        
        <Divider 
          sx={{ 
            width: '100%', 
            color: '#828282',
            fontSize: '10px', 
            mt: '7px'
          }}
        >
          or continue with
        </Divider>

        <GoogleLogin
          onSuccess={handleGoogleLoginSuccess}
          onError={handleGoogleLoginError}
          useOneTap={false}
          theme='outline'
          size='medium'
          text='continue_with'
          width='100%'
        />
        
        <div style={{
          color: '#828282',
          fontSize: '10px',
          marginTop: '5px',
          textAlign: 'center'
        }}>
          By clicking continue, you agree to our&nbsp;
          <a href="https://policies.google.com/terms?hl=en" style={{ color: '#000000' }}>Terms of Service</a>
            &nbsp;
            and&nbsp;
            <a href="https://policies.google.com/privacy?hl=en" style={{ color: '#000000' }}>Privacy Policy</a>
        </div>

      </Box>
    </Box>
  );
};

export default SignUpForm;
