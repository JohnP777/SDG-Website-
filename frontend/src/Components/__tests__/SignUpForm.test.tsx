import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import SignUpForm from "../SignUpForm";
import * as ApiCalls from "../../Utilities/ApiCalls"
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { GoogleOAuthProvider } from '@react-oauth/google';

// Mock the API calls
jest.mock("../../Utilities/ApiCalls", () => ({
  apiCallPost: jest.fn(),
}));

// Mock window.location.reload
const mockReload = jest.fn();
Object.defineProperty(window, 'location', {
  value: {
    reload: mockReload,
  },
  writable: true
});

// Mock useNavigate
const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

// Helper function to render the component with router
const renderWithRouter = (component: React.ReactNode) => {
  return render(
    <GoogleOAuthProvider clientId="google-client-id">
      <MemoryRouter>
        {component}
      </MemoryRouter>
    </GoogleOAuthProvider>
  );
};

describe("CreateSignUpForm", () => {
  // Reset mocks before each test
  beforeEach(() => {
    jest.clearAllMocks();
    (ApiCalls.apiCallPost as jest.Mock).mockResolvedValue({
      statusCode: 200
    });
  });

  // Snapshot test
  it("should match snapshot", async () => {
    const { asFragment } = renderWithRouter(<SignUpForm />);
    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });

  // Test sign up input
  it("should update sign up when input changes", () => {
    renderWithRouter(<SignUpForm />);
    
    // Find the input and change its value
    const usernameInput = screen.getByPlaceholderText("Username");
    const emailInput = screen.getByPlaceholderText("Email");
    const passwordInput = screen.getByPlaceholderText("Password");
    const confirmPasswordInput = screen.getByPlaceholderText("Confirm Password");
    fireEvent.change(usernameInput, { target: { value: "testuser" } });
    fireEvent.change(emailInput, { target: { value: "testemail@gmail.com" } });
    fireEvent.change(passwordInput, { target: { value: "TestPassword1234!" } });
    fireEvent.change(confirmPasswordInput, { target: { value: "TestPassword1234!" } });
    
    // Check if the input value has been updated
    expect(usernameInput).toHaveValue("testuser");
    expect(emailInput).toHaveValue("testemail@gmail.com");
    expect(passwordInput).toHaveValue("TestPassword1234!");
    expect(confirmPasswordInput).toHaveValue("TestPassword1234!");
  });

  // Test user sign up functionality
  it("should allow the user to sign up", async () => {

    jest.clearAllMocks();
    (ApiCalls.apiCallPost as jest.Mock).mockImplementation((url) => {
      if (url === 'api/auth/pending-register/') {
        return Promise.resolve({
          statusCode: 400,
          username: ["This field may not be blank."],
          email: ["This field may not be blank."],
          password1: ["This field may not be blank."],
          password2: ["This field may not be blank."]
        });
      }
    });

    renderWithRouter(<SignUpForm />);
  
    const signUpButton = screen.getByText("Sign Up");

    userEvent.click(signUpButton);

    // Error should show up (since no inputs are fed in)
    await waitFor(() => {
      expect(screen.getByText("USERNAME: This field may not be blank.")).toBeInTheDocument();
      expect(screen.getByText("EMAIL: This field may not be blank.")).toBeInTheDocument();
      expect(screen.getByText("PASSWORD1: This field may not be blank.")).toBeInTheDocument();
      expect(screen.getByText("PASSWORD2: This field may not be blank.")).toBeInTheDocument();
    });
  
  })  
  
  it("on success, should navigate to confirmation pin page", async () => {
    jest.clearAllMocks();
    (ApiCalls.apiCallPost as jest.Mock).mockImplementation((url) => {
      if (url === 'api/auth/pending-register/') {
        return Promise.resolve({
          statusCode: 200,
          token: "fakeToken",
          message: "Verification code sent. Please check your email."
        });
      }
    });

    renderWithRouter(<SignUpForm />);

    // Check inputs are there
    const usernameInput = screen.getByPlaceholderText("Username");
    const emailInput = screen.getByPlaceholderText("Email");
    const passwordInput = screen.getByPlaceholderText("Password");
    const confirmPasswordInput = screen.getByPlaceholderText("Confirm Password");
    const signUpButton = screen.getByText("Sign Up");

    fireEvent.change(usernameInput, { target: { value: "testuser" } });
    fireEvent.change(emailInput, { target: { value: "testemail@gmail.com" } });
    fireEvent.change(passwordInput, { target: { value: "TestPassword1234!" } });
    fireEvent.change(confirmPasswordInput, { target: { value: "TestPassword1234!" } });
  
    // Test form submission success
    userEvent.click(signUpButton);

    await waitFor(() => {
      expect(mockedUsedNavigate).toHaveBeenCalledWith(
        '/confirmation-pin', 
        expect.objectContaining({
          state: expect.objectContaining({
            username: "testuser",
            email: "testemail@gmail.com",
            password1: "TestPassword1234!",
            password2: "TestPassword1234!",
            mobile: "",
            initialToken: "fakeToken"
          })
        })
      )
    });
  })
});