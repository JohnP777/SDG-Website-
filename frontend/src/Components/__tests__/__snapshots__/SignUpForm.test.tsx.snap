// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateSignUpForm should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1gaussg"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
      style="font-size: 1.6rem; font-weight: bold; margin-bottom: 20px; text-align: center;"
    >
      SDG Knowledge Sign Up
    </p>
    <div
      class="MuiBox-root css-1x1nri4"
    >
      <input
        name="username"
        placeholder="Username"
        style="width: 85%; padding: 5px 10px; font-size: 12px; border: 1px solid #e0e0e0; border-radius: 5px; box-sizing: border-box;"
        type="text"
        value=""
      />
      <input
        name="contact"
        placeholder="Email"
        style="width: 85%; padding: 5px 10px; font-size: 12px; border: 1px solid #e0e0e0; border-radius: 5px; box-sizing: border-box;"
        type="text"
        value=""
      />
      <input
        name="phonenumber"
        placeholder="Phone Number (optional)"
        style="width: 85%; padding: 5px 10px; font-size: 12px; border: 1px solid #e0e0e0; border-radius: 5px; box-sizing: border-box;"
        type="text"
        value=""
      />
      <input
        name="password"
        placeholder="Password"
        style="width: 85%; padding: 5px 10px; font-size: 12px; border: 1px solid #e0e0e0; border-radius: 5px; box-sizing: border-box;"
        type="password"
        value=""
      />
      <input
        name="confirmpassword"
        placeholder="Confirm Password"
        style="width: 85%; padding: 5px 10px; font-size: 12px; border: 1px solid #e0e0e0; border-radius: 5px; box-sizing: border-box;"
        type="password"
        value=""
      />
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-1ejnzjp-MuiFormControlLabel-root"
      >
        <span
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeSmall PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeSmall MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeSmall css-imrjgg-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            class="PrivateSwitchBase-input css-j8yymo"
            data-indeterminate="false"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
            data-testid="CheckBoxOutlineBlankIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-rizt0-MuiTypography-root"
        >
          Show Password
        </span>
      </label>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorPrimary css-1eioe9r-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Sign Up
      </button>
      <div
        aria-orientation="horizontal"
        class="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-jyffef-MuiDivider-root"
        role="separator"
      >
        <span
          class="MuiDivider-wrapper css-1134oje-MuiDivider-wrapper"
        >
          or continue with
        </span>
      </div>
      <div
        style="height: 32px;"
      />
      <div
        style="color: rgb(130, 130, 130); font-size: 10px; margin-top: 5px; text-align: center;"
      >
        By clicking continue, you agree to our 
        <a
          href="https://policies.google.com/terms?hl=en"
          style="color: rgb(0, 0, 0);"
        >
          Terms of Service
        </a>
          and 
        <a
          href="https://policies.google.com/privacy?hl=en"
          style="color: rgb(0, 0, 0);"
        >
          Privacy Policy
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
